File: algorithms/leetcode/013-roman-to-integer/roman-to-integer.js
const romanToInt = s => {
    const rome = new Map();
    rome.set('M', 1000);
    rome.set('D', 500);
    rome.set('C', 100);
    rome.set('L', 50);
    rome.set('X', 10);
    rome.set('V', 5);
    rome.set('I', 1);

    let result = 0;

    for (let i = 0; i < s.length - 1; i++) {
        const cur = rome.get(s[i]);
        const next = rome.get(s[i + 1]);

        if (next && next > cur) {
            result -= cur;
        } else {
            result += cur;
        }
    }

    return result + rome.get(s[s.length - 1]);
};

File: leetcode/013-roman-to-integer/RomanToInteger.swift
//
//  RomanToInteger.swift
//  LeetCode
//
//  Created by Valerii Iatsko on 10/7/20.
//

import Foundation

class Solution {
    func romanToInt(_ s: String) -> Int {
        let dict: [Character: Int] = [
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000
        ]

        var result = 0
        var prev = -1

        for ch in s.reversed() {
            if let curr = dict[ch] {
                if (curr >= prev) {
                    result += curr
                } else {
                    result -= curr
                }

                prev = curr
            }
        }

        return result
    }
}
